### ISSUE TITLE
Review impact of extended property patterns on existing rules

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#3984]
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/extended-property-patterns

**Comment by sharwell** [comment]
Audit by Deep Research

# Impact of Extended Property Patterns on StyleCopAnalyzers 

Extended property patterns (introduced in C# 10) allow pattern matching directly on nested properties using dot notation (for example, `obj is { Inner.Value: 0 }` instead of `obj is { Inner: { Value: 0 } }`). This new syntax impacts several StyleCopAnalyzers rules, especially those dealing with member access and spacing in patterns.

This checklist covers each impacted rule, how extended property patterns interact with it, and what is required for full support.

## SA1101 – Prefix local calls with `this`

### Interaction

SA1101 requires instance member accesses to be prefixed with `this.` for clarity. Extended property patterns caused false positives because the analyzer treated nested property references in the pattern as instance members of the current class instead of members of the pattern input.

Example:

```csharp
return arg is { Inner.Value: "" };
```

Earlier, SA1101 incorrectly flagged `Inner.Value` and suggested `this.Inner.Value`, which is wrong. In this context, `Inner.Value` is a property of `arg` (the pattern input), not of `this`.

### Required changes

* **Implementation**

  Update SA1101 to recognize property pattern contexts and exclude them from the “must be prefixed with `this`” rule.

  Specifically:

  * When analyzing an extended property pattern, *the expression to the left of the `:` in a subpattern* must not be required (or allowed) to use `this.`.
  * The analyzer should treat these as pattern members of the input expression, not as members of the containing type.

* **Tests**

  Add or extend tests to ensure:

  * Patterns like:

    ```csharp
    return obj is { Prop.SubProp: value };
    ```

    do **not** produce SA1101 diagnostics.

  * Genuine missing `this.` in non-pattern contexts are still reported.

  * Pattern usage in different contexts is covered:

    * `is` expressions:

      ```csharp
      if (item is { Child.Name: "X" })
      {
      }
      ```

      No SA1101 diagnostic on `Child.Name`.

    * `switch` statements and expressions:

      ```csharp
      return arg switch
      {
          { Inner.Value: > 0 } => 1,
          _ => 0,
      };
      ```

    * Combinations with other patterns (`and`, `or`, parenthesized patterns) to ensure SA1101 ignores *all* pattern contexts.

* **Documentation**

  Update the SA1101 documentation to:

  * State that property pattern syntax is excluded from the rule.
  * Include an example like:

    ```csharp
    if (item is { Outer.InnerProp: value })
    {
    }
    ```

    and explicitly note that `Outer.InnerProp` is valid without `this.` in this context.

### Test scenarios to cover

* `if (item is { Child.Name: "X" })` – no SA1101 diagnostic.
* `case MyType { Prop.Value: > 0 }: ...` – no SA1101 diagnostic.
* Nested property patterns inside:

  * `switch` expressions,
  * `switch` statements,
  * patterns combined with `and` / `or`,
  * parenthesized patterns.

All should be free of SA1101 diagnostics as long as the access occurs in the property pattern portion of the match.

---

## SA1024 – ColonsMustBeSpacedCorrectly

### Interaction

SA1024 governs spacing around `:` tokens. Extended property patterns introduce a new colon usage: separating a nested property path from the subpattern or constant, for example:

```csharp
obj is { Inner.Value: "" }
```

The expected style is:

* **No space** before the colon.
* **Single space** after the colon (in most situations), e.g.:

```csharp
obj is { Inner.Value: "" }
```

Originally SA1024 did not account for colons appearing in the new pattern grammar and could:

* Ignore spacing around these colons, or
* Apply spacing rules from unrelated contexts.

### Required changes

* **Implementation**

  Extend SA1024’s logic to recognize colons inside extended property patterns:

  * In Roslyn, the colon in a property subpattern can appear in a `SubpatternSyntax` (rather than just `NameColonSyntax`).
  * When the colon’s parent is a pattern clause or `SubpatternSyntax`, SA1024 should:

    * Forbid a space before the colon.
    * Generally require a single space after the colon (consistent with other named patterns and colon usage in the rule).

* **Tests**

  Add or extend tests to cover:

  * Correct spacing:

    ```csharp
    return x is { Prop.Value: 1 };
    ```

    No SA1024 diagnostics.

  * Space **before** colon:

    ```csharp
    return x is { Prop.Value : 1 };
    ```

    Expect SA1024 (“Colon should not be preceded by a space”).

  * Missing space **after** colon:

    ```csharp
    return x is { Prop.Value:1 };
    ```

    Expect SA1024 (“Colon should be followed by a space”).

  * Single-level vs multi-level property names:

    ```csharp
    obj is { Prop: val }
    obj is { Prop.SubProp: val }
    ```

  * Multiple subpatterns:

    ```csharp
    obj is { X.Y: 1, Z: 2 }
    ```

    Each colon must be checked independently.

* **Documentation**

  Update SA1024 documentation to:

  * Explicitly call out property patterns, for example:

    > In patterns (e.g. `obj is { Property.SubProperty: value }`), the colon is treated like a named element separator: no space before, and one space after.

  * Include at least one sample showing compliant and non-compliant property pattern spacing.

### Test scenarios to cover

* Correctly spaced pattern:

  ```csharp
  if (o is { A.B: 5 })
  {
  }
  ```

  No SA1024 diagnostic.

* Mis-spaced patterns:

  ```csharp
  if (o is { A.B : 5 }) { } // space before colon
  if (o is { A.B:5 }) { }   // missing space after colon
  ```

  Both should produce SA1024 diagnostics.

* Mixed pattern with multiple subpatterns:

  ```csharp
  if (o is { X.Y: 1, Z: 2 }) { }
  ```

  All colons should be evaluated according to the rule.

---

## Other rules

Other StyleCopAnalyzers rules were evaluated and found not to be meaningfully affected by extended property patterns. General pattern-matching support (introduced in earlier C# versions) already covered most relevant syntax:

* Spacing around `is` keyword.
* Braces in patterns.
* Other pattern shapes.

After implementing the changes to SA1101 and SA1024 (with tests and documentation updates), StyleCopAnalyzers should correctly handle extended property pattern syntax without false positives or missed violations.

### Sources

* `SA1101CSharp10UnitTests.cs`
* Issue **“SA1101 in property pattern”** (`#3472`)
* StyleCopAnalyzers releases
* `SA1024ColonsMustBeSpacedCorrectly.cs`
